apache-rat-core/src/main/java/org/apache/rat/document/DocumentNameMatcher.java

```diff
@@ -219,4 +314,225 @@ public static DocumentNameMatcher and(final Collection<DocumentNameMatcher> matc
+     */
+    abstract static class CollectionPredicateImpl implements CollectionPredicate {
+        /** The collection for matchers that make up this predicate */
+        private final Iterable<DocumentNameMatcher> matchers;
+
+        /**
+         * Constructs a collecton predicate from the collection of matchers
+         * @param matchers the colleciton of matchers to use.
+         */
+        protected CollectionPredicateImpl(final Iterable<DocumentNameMatcher> matchers) {
+            this.matchers = matchers;
+        }
+
+        /**
+         * Gets the internal matchers.
+         * @return an iterable over the internal matchers.
+         */
+        public Iterable<DocumentNameMatcher> getMatchers() {
+            return matchers;
+        }
+
+        public String toString() {
+            StringBuilder builder = new StringBuilder(this.getClass().getName()).append(": ").append(System.lineSeparator());
+            for (DocumentNameMatcher matcher : matchers) {
+                builder.append(matcher.predicate.toString()).append(System.lineSeparator());
+            }
+            return builder.toString();
+        }
+    }
+
+    /**
+     * An implementation of "and" logic across a collection of DocumentNameMatchers.
+     */
+    // package private for testing access
+    static class And extends CollectionPredicateImpl {
+        And(final Iterable<DocumentNameMatcher> matchers) {
+            super(matchers);
+        }
+
+        @Override
+        public boolean test(final DocumentName documentName) {
+            for (DocumentNameMatcher matcher : getMatchers()) {
+                if (!matcher.matches(documentName)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    /**
+     * An implementation of "or" logic across a collection of DocumentNameMatchers.
+     */
+    // package private for testing access
+    static class Or extends CollectionPredicateImpl {
+        Or(final Iterable<DocumentNameMatcher> matchers) {
+            super(matchers);
+        }
+
+        @Override
+        public boolean test(final DocumentName documentName) {
+            for (DocumentNameMatcher matcher : getMatchers()) {
+                if (matcher.matches(documentName)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    /**
+     * An implementation of "or" logic across a collection of DocumentNameMatchers.
+     */
+    // package private for testing access
+    static class MatcherPredicate extends CollectionPredicateImpl {
+        MatcherPredicate(final Iterable<DocumentNameMatcher> matchers) {
+            super(matchers);
+        }
+
+        @Override
+        public boolean test(final DocumentName documentName) {
+            Iterator<DocumentNameMatcher> iter = getMatchers().iterator();
+            // included
+            if (iter.next().matches(documentName)) {
```

Comment:
Is it correct to call iter.next() twice here or does that unintentionally skip the underlying iterator twice instead of a stepwise iteration OR is this a performance trick I just did not know about to iterate in a faster fashion?